File: Java/Daily/Medium/AvoidFlood.java
import java.util.*;
// 1488
class Solution {
    public int[] avoidFlood(int[] rains) {
        int n = rains.length;
        int[] res = new int[n];
        Arrays.fill(res, -1);
        Map<Integer, Integer> full = new HashMap<>();
        TreeSet<Integer> dryDays = new TreeSet<>();

        for (int i = 0; i < n; i++) {
            int lake = rains[i];

            if (lake == 0) {
                dryDays.add(i);
                res[i] = 1; // placeholder
            } else {
                if (full.containsKey(lake)) {
                    int lastFilled = full.get(lake);
                    Integer dryDay = dryDays.ceiling(lastFilled + 1);
                    if (dryDay == null) return new int[0]; // no valid dry day
                    res[dryDay] = lake;
                    dryDays.remove(dryDay);
                }
                full.put(lake, i);
                res[i] = -1;
            }
        }

        return res;
    }
}

File: Daily/AvoidFlood.py
from typing import List
from collections import defaultdict
# 1488
class Solution:
    def avoidFlood(self, rains: List[int]) -> List[int]:
        
        full = defaultdict(int)
        res = []
        dry = []
        for r in range(len(rains)):
            if rains[r] > 0:
                if rains[r] not in full:
                    full[rains[r]] = r
                    res.append(-1)
                else:
                    x = full[rains[r]]
                    if dry:
                        s = dry[0]
                        if x < s < r:
                            res.append(s)
                            full[rains[r]] = r
                            dry.pop(0)
                    else:
                        return []
            else:
                dry.append(r)
        return res

                
        